README
-----------------------------------------------------------------------------
| PadWalker v1.91    - Robin Houston
-----------------------------------------------------------------------------

NAME
       PadWalker - play with other peoples' lexical variables

SYNOPSIS
	 use PadWalker qw(peek_my peek_our peek_sub closed_over);
	 ...

DESCRIPTION
       PadWalker is a module which allows you to inspect (and even change!)
       lexical variables in any subroutine which called you. It will only show
       those variables which are in scope at the point of the call.

       PadWalker is particularly useful for debugging. It's even used by
       Perl's built-in debugger. (It can also be used for evil, of course.)

       I wouldn't recommend using PadWalker directly in production code, but
       it's your call. Some of the modules that use PadWalker internally are
       certainly safe for and useful in production.

       peek_my LEVEL
       peek_our LEVEL
	   The LEVEL argument is interpreted just like the argument to
	   "caller".  So peek_my(0) returns a reference to a hash of all the
	   "my" variables that are currently in scope; peek_my(1) returns a
	   reference to a hash of all the "my" variables that are in scope at
	   the point where the current sub was called, and so on.

	   "peek_our" works in the same way, except that it lists the "our"
	   variables rather than the "my" variables.

	   The hash associates each variable name with a reference to its
	   value. The variable names include the sigil, so the variable $x is
	   represented by the string '$x'.

	   For example:

	     my $x = 12;
	     my $h = peek_my (0);
	     ${$h->{'$x'}}++;

	     print $x;	# prints 13

	   Or a more complex example:

	     sub increment_my_x {
	       my $h = peek_my (1);
	       ${$h->{'$x'}}++;
	     }

	     my $x=5;
	     increment_my_x;
	     print $x;	# prints 6

       peek_sub SUB
	   The "peek_sub" routine takes a coderef as its argument, and returns
	   a hash of the "my" variables used in that sub. The values will usu-
	   ally be undefined unless the sub is in use (i.e. in the call-chain)
	   at the time. On the other hand:

	     my $x = "Hello!";
	     my $r = peek_sub(sub {$x})->{'$x'};
	     print "$$r\n";	   # prints 'Hello!'

	   If the sub defines several "my" variables with the same name,
	   you'll get the last one. I don't know of any use for "peek_sub"
	   that isn't broken as a result of this, and it will probably be dep-
	   recated in a future version in favour of some alternative inter-
	   face.

       closed_over SUB
	   "closed_over" is similar to "peek_sub", except that it only lists
	   the "my" variables which are used in the subroutine but defined
	   outside: in other words, the variables which it closes over. This
	   does have reasonable uses: see Data::Dump::Streamer, for example (a
	   future version of which may in fact use "closed_over").

       set_closed_over SUB, HASH_REF
	   "set_closed_over" reassigns the pad variables that are closed over
	   by the subroutine.

	   The second argument is a hash of references, much like the one
	   returned from "closed_over".

       var_name LEVEL, VAR_REF
       var_name SUB,   VAR_REF
	   "var_name(sub, var_ref)" returns the name of the variable referred
	   to by "var_ref", provided it is a "my" variable used in the sub.
	   The "sub" parameter can be either a CODE reference or a number. If
	   it's a number, it's treated the same way as the argument to
	   "peek_my".

	   For example,

	     my $foo;
	     print var_name(0, \$foo);	  # prints '$foo'

	     sub my_name {
	       return var_name(1, shift);
	     }
	     print my_name(\$foo);	  # ditto

AUTHOR
       Robin Houston <robin@cpan.org>

       With contributions from Richard Soberberg, Florian Ragwitz and Yuval Kogman,
       bug-spotting from Peter Scott and Dave Mitchell, and suggestions from demerphq.

SEE ALSO
       Devel::LexAlias, Devel::Caller, Sub::Parameters

COPYRIGHT
       Copyright (c) 2000-2010, Robin Houston. All Rights Reserved.  This mod-
       ule is free software. It may be used, redistributed and/or modified
       under the same terms as Perl itself.


Changes
Revision history for Perl extension PadWalker.

0.01  Thu Nov  9 12:58:10 2000
	- original version; created by h2xs 1.19

Revision history between 0.01 and 0.03 has been lost in the mists
of time. Sorry about that.

0.03 was the first public release.

0.04  Thu Jul 19 13:50:19 BST 2001
	- Applied patch from Richard Soderberg to let it compile
	  under ithreads

0.05  Thu Jan 10 21:12:10 GMT 2002
	- Experimental peek_sub routine

0.06  Wed Mar  6 22:16:13 GMT 2002
	- Proper test script
	- Two bug fixes

0.07  Thu Mar 14 19:56:29 GMT 2002
	- Clean up POD documentation
	- Behave properly with debugger
	- Work under Perl 5.005

0.08  Mon Mar 18 17:54:16 GMT 2002
	- _upcontext() XSUB, for Richard Clamp
	- work properly with recursion
	- see past eval

0.09  Wed May 21 17:24:32 BST 2003
        - compile on threaded builds
        - skip test 9 if we're on Perl 5.8. (Otherwise it fails.)

0.10  Wed Jul 30 18:40:03 BST 2003
	- accommodate the new versions of perl (5.8.1)
	- acknowledge that test 9 fails because nested eval handling
          is simply broken. (Fix in a future release.)

0.11  Wed Aug 10 20:06:41 BST 2005 
	- fix various nasty bugs, specifically:
	 o peek_my(1) now works correctly in a sub called from the top level;
	 o deal better with sub calls across source files
	- don't return 'our' variables. (It is peek_my(), after all!
	  And the values were never properly returned.)

0.12  Tue Aug 16 15:40:04 BST 2005
	- make it work (up to a point) with Perl 5.6.

0.13  Mon Oct  3 11:54:23 BST 2005
	- don't build a debugging build unless explicitly told to!

0.14  Thu Oct  6 17:19:06 BST 2005
	- Fix the bugs reported by Dave Mitchell:
	 o if one variable masks another, make sure we return the
	   appropriate one;
	 o for a variable whose value has been lost, return undef
	   rather than the name of the variable;
	 o Don't die if PadWalker is called from a closure whose
           containing scope has disappeared.

0.99  Fri Oct  7 17:23:09 BST 2005
	- Make peek_sub return the values, if possible,
 	  even when it's not in the call chain;
	- Allow an our variable to mask a my variable,
	  and vice versa;
	- Add peek_our and closed_over routines.

0.99_91  Thu Oct 13 17:35:11 BST 2005
	- Make peek_my work correctly when called from a do "file";
	- Add var_name routine;
	- Add an :all EXPORT_TAG;
	- Drop support for perl < 5.8;
	- Eliminate compiler warnings (at least on gcc).

0.99_92  Sat Oct 15 00:17:24 BST 2005
	- Add license info to META.yml;
	- Use tabs consistently in the Changes file...
	- Don't use snprintf (apparently Bill Gates doesn't approve);
	- ignore 'our' variables in closed_over();
	- Eliminate more compiler warnings, this time on Win32.

0.99_93  Fri Oct 28 13:18:20 BST 2005
	- Change the sekrit undocumented second return value
	  of closed_over() to something slightly different.

1.0  Wed Nov  2 12:25:49 GMT 2005
	- Bump version number to 1.0

1.1  Sun Oct 22 16:13:40 BST 2006
	- Accommodate change 27312 "Store the stash for our in magic slot"
	  (See http://www.mail-archive.com/perl5-changes@perl.org/msg14073.html
	  or http://public.activestate.com/cgi-bin/perlbrowse/27312)
	
	  See also http://public.activestate.com/cgi-bin/perlbrowse/27306

1.2  Thu Nov 16 22:33:27 GMT 2006
	- Change prerequisites to accurately reflect versions of Perl that
	  PadWalker actually works with (i.e. 5.8.2 or later).
	- Fix memory leak: thanks to Rocco Caputo

1.3  Tue Jan  2 23:10:35 GMT 2007
	- Accommodate changes 29629-29630 "Move the low/high cop sequences
	  from NVX/IVX to a two U32 structure".

1.4  Fri Jan  5 09:12:11 GMT 2007
	- Accommodate change 29679 "Rename OURSTASH to SvOURSTASH and
	  OURSTASH_set to SvOURSTASH_set". (Dear Nick, please stop
	  breaking PadWalker. kthxbye.)

1.5  Fri Jan  5 16:22:27 GMT 2007
	- Fix egregrious bug in 1.4 :-(

1.6  Mon Jan 14 10:48:09 GMT 2008
	- Make _upcontext work in 64-bit architectures.
	  (http://rt.cpan.org/Ticket/Display.html?id=32287)
	  Thanks to Niko Tyni.

1.7  Mon Feb  4 09:56:31 GMT 2008
	- Keep up with changes in blead post-5.10 (@33030)

1.8  Thu 25 Jun 2009 21:17:17 BST
	- Apply patches from doy (#41710) and nothingmuch (set_closed_over).

1.9  Fri 26 Jun 2009 10:01:17 BST
	- Identical to 1.8, but with the bogus metadata ._ files removed
	  from the distributed tar file.

1.91  Wed 14 Jul 2010 01:07:05 BST
	- Incorporate patches from Florian Ragwitz and Yuval Kogman
          (see http://github.com/robinhouston/PadWalker/commits/master)



Makefile.PL
use ExtUtils::MakeMaker;

use strict;
require 5.008002;

# Remember (like I didn't) that WriteMakefile looks at @ARGV,
# so an alternative way to configure a debugging build is:
# perl Makefile.PL DEFINE=-DPADWALKER_DEBUGGING.
my $DEBUGGING = '';
if (@ARGV && $ARGV[0] eq '-d') {
  warn "Configuring a debugging build of PadWalker\n";
  print STDERR <<END;
************************************************************************
* WARNING!  WARNING!  WARNING!  WARNING!  WARNING!  WARNING!  WARNING! *
************************************************************************
You are building PadWalker in debugging mode, which causes it to
print a lot of gnomic information about its internal operation.
The test suite will fail, because this information will confuse
the test harness. You almost certainly do *not* want to do this
unless you're the author of PadWalker (or perhaps just irrepressibly
curious about its internal operation).

END

  $DEBUGGING = '-DPADWALKER_DEBUGGING';
  shift;
}

WriteMakefile(
    'NAME'	=> 'PadWalker',
    'VERSION_FROM' => 'PadWalker.pm', # finds $VERSION
    'LIBS'	=> [''],   # e.g., '-lm' 
    'DEFINE'	=> $DEBUGGING,
    'INC'	=> '',     # e.g., '-I/usr/include/other',
    ($DEBUGGING ? (CCFLAGS => '-Wall -ansi') : ()),
    'NO_META'   => 1,
);


PadWalker.c
/*
 * This file was generated automatically by xsubpp version 1.9508 from the
 * contents of PadWalker.xs. Do not edit this file, edit PadWalker.xs instead.
 *
 *	ANY CHANGES MADE HERE WILL BE LOST!
 *
 */

#line 1 "PadWalker.xs"
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

/* For development testing */
#ifdef PADWALKER_DEBUGGING
# define debug_print(x) printf x
#else
# define debug_print(x)
#endif

/* For 5.005 compatibility */
#ifndef aTHX_
#  define aTHX_
#endif
#ifndef pTHX_
#  define pTHX_
#endif
#ifndef pTHX
#  define pTHX
#endif
#ifndef aTHX
#  define aTHX
#endif
#ifndef CxTYPE
#  define CxTYPE(cx) ((cx)->cx_type)
#endif

/* For debugging */
#ifdef PADWALKER_DEBUGGING
char *
cxtype_name(U32 cx_type)
{
  switch(cx_type & CXTYPEMASK)
  {
    case CXt_NULL:   return "null";
    case CXt_SUB:    return "sub";
    case CXt_EVAL:   return "eval";
    case CXt_LOOP:   return "loop";
    case CXt_SUBST:  return "subst";
    case CXt_BLOCK:  return "block";
    case CXt_FORMAT: return "format";

    default:         debug_print(("Unknown context type 0x%x\n", cx_type));
					 return "(unknown)";
  }
}

void
show_cxstack(void)
{
	I32 i;
    for (i = cxstack_ix; i>=0; --i)
    {
		printf(" =%ld= %s (%x)", (long)i,
				cxtype_name(CxTYPE(&cxstack[i])), cxstack[i].blk_oldcop->cop_seq);
        if (CxTYPE(&cxstack[i]) == CXt_SUB) {
		  CV *cv = cxstack[i].blk_sub.cv;
		  printf("\t%s", (cv && CvGV(cv)) ? GvNAME(CvGV(cv)) :"(null)");
		}
		putchar('\n');
    }
}
#else
# define show_cxstack()
#endif

/* Originally stolen from pp_ctl.c; now significantly different */

I32
dopoptosub_at(pTHX_ PERL_CONTEXT *cxstk, I32 startingblock)
{
    dTHR;
    I32 i;
    PERL_CONTEXT *cx;
    for (i = startingblock; i >= 0; i--) {
        cx = &cxstk[i];
        switch (CxTYPE(cx)) {
        default:
            continue;
        case CXt_SUB:
    	/* In Perl 5.005, formats just used CXt_SUB */
#ifdef CXt_FORMAT
       case CXt_FORMAT:
#endif
            debug_print(("**dopoptosub_at: found sub #%ld\n", (long)i));
            return i;
        }
    }
	debug_print(("**dopoptosub_at: not found #%ld\n", (long)i));
    return i;
}

I32
dopoptosub(pTHX_ I32 startingblock)
{
    dTHR;
    return dopoptosub_at(aTHX_ cxstack, startingblock);
}

/* This function is based on the code of pp_caller */
PERL_CONTEXT*
upcontext(pTHX_ I32 count, COP **cop_p, PERL_CONTEXT **ccstack_p,
				I32 *cxix_from_p, I32 *cxix_to_p)
{
    PERL_SI *top_si = PL_curstackinfo;
    I32 cxix = dopoptosub(aTHX_ cxstack_ix);
    PERL_CONTEXT *cx;
    PERL_CONTEXT *ccstack = cxstack;
    I32 dbcxix;

	if (cxix_from_p) *cxix_from_p = cxstack_ix+1;
	if (cxix_to_p)   *cxix_to_p   = cxix;
    for (;;) {
        /* we may be in a higher stacklevel, so dig down deeper */
        while (cxix < 0 && top_si->si_type != PERLSI_MAIN) {
            top_si  = top_si->si_prev;
            ccstack = top_si->si_cxstack;
            cxix = dopoptosub_at(aTHX_ ccstack, top_si->si_cxix);
			if (cxix_to_p && cxix_from_p) *cxix_from_p = *cxix_to_p;
			if (cxix_to_p) *cxix_to_p = cxix;
        }
        if (cxix < 0 && count == 0) {
		    if (ccstack_p) *ccstack_p = ccstack;
            return (PERL_CONTEXT *)0;
		}
        else if (cxix < 0)
            return (PERL_CONTEXT *)-1;
        if (PL_DBsub && cxix >= 0 &&
                ccstack[cxix].blk_sub.cv == GvCV(PL_DBsub))
            count++;
        if (!count--)
            break;

        if (cop_p) *cop_p = ccstack[cxix].blk_oldcop;
        cxix = dopoptosub_at(aTHX_ ccstack, cxix - 1);
			if (cxix_to_p && cxix_from_p) *cxix_from_p = *cxix_to_p;
			if (cxix_to_p) *cxix_to_p = cxix;
    }
    if (ccstack_p) *ccstack_p = ccstack;
    return &ccstack[cxix];
}

/* end thievery */

void
pads_into_hash(AV* pad_namelist, AV* pad_vallist, HV* hash, U32 valid_at_seq)
{
    I32 i;

    for (i=0; i<=av_len(pad_namelist); ++i) {
      SV** name_ptr = av_fetch(pad_namelist, i, 0);

      if (name_ptr) {
        SV*   name_sv = *name_ptr;

	if (SvPOKp(name_sv)) {
          char* name_str = SvPVX(name_sv);

        debug_print(("** %s (%x,%x) [%x]%s\n", name_str,
               I_32(SvNVX(name_sv)), SvIVX(name_sv), valid_at_seq,
               SvFAKE(name_sv) ? " <fake>" : ""));
        
        /* Check that this variable is valid at the cop_seq
         * specified, by peeking into the NV and IV slots
         * of the name sv. (This must be one of those "breathtaking
         * optimisations" mentioned in the Panther book).

         * Anonymous subs are stored here with a name of "&",
         * so also check that the name is longer than one char.
         * (Note that the prefix letter is here as well, so a
         * valid variable will _always_ be >1 char)

		 * We ignore 'our' variables, since you can always dig
		 * them out of the stash directly.
         */

        if (!(SvFLAGS(name_sv) & SVpad_OUR) &&
		    (SvFAKE(name_sv) || 0 == valid_at_seq ||
            (valid_at_seq <= SvIVX(name_sv) &&
            valid_at_seq > I_32(SvNVX(name_sv)))) &&
            strlen(name_str) > 1 )

          {
            SV **val_ptr, *val_sv;

            val_ptr = av_fetch(pad_vallist, i, 0);
            val_sv = val_ptr ? *val_ptr : &PL_sv_undef;
			
	        hv_store(hash, name_str, strlen(name_str),
                     newRV_inc(val_sv), 0);
          }
        }
      }
    }
}

void
padlist_into_hash(AV* padlist, HV* hash, U32 valid_at_seq, U16 depth)
{
    /* We blindly deref this, cos it's always there (AFAIK!) */
    AV* pad_namelist = (AV*) *av_fetch(padlist, 0, FALSE);
    AV* pad_vallist  = (AV*) *av_fetch(padlist, depth, FALSE);

    pads_into_hash(pad_namelist, pad_vallist, hash, valid_at_seq);
}

void
context_vars(PERL_CONTEXT *cx, HV* ret, U32 seq, CV *cv)
{
    /* If cx is null, we take that to mean that we should look
     * at the cv instead
     */

	debug_print(("**context_vars(%p, %p, 0x%lx)\n",
			(void*)cx, (void*)ret, (long)seq));
    if (cx == (PERL_CONTEXT*)-1)
        croak("Not nested deeply enough");

    else {
        CV* cur_cv = cx ? cx->blk_sub.cv           : cv;
        U16 depth  = cx ? cx->blk_sub.olddepth + 1 : 1;

        if (!cur_cv)
            die("panic: Context has no CV!\n");
    
        while (cur_cv) {
            debug_print(("\tcv name = %s; depth=%d\n",
                    CvGV(cur_cv) ? GvNAME(CvGV(cur_cv)) :"(null)", depth));
            padlist_into_hash(CvPADLIST(cur_cv), ret, seq, depth);
            cur_cv = CvOUTSIDE(cur_cv);
            if (cur_cv) depth  = CvDEPTH(cur_cv);
        }
    }
}

#line 247 "PadWalker.c"
XS(XS_PadWalker_peek_my); /* prototype to pass -Wmissing-prototypes */
XS(XS_PadWalker_peek_my)
{
    dXSARGS;
    if (items != 1)
	Perl_croak(aTHX_ "Usage: PadWalker::peek_my(uplevel)");
    SP -= items;
    {
	I32	uplevel = (I32)SvIV(ST(0));
#line 244 "PadWalker.xs"
    HV* ret = newHV();
    PERL_CONTEXT *cx, *ccstack;
    COP *cop = 0;
    I32 cxix_from, cxix_to, i;

#line 263 "PadWalker.c"
#line 250 "PadWalker.xs"
    show_cxstack();
    if (PL_curstackinfo->si_type != PERLSI_MAIN)
	  debug_print(("!! We're in a higher stack level\n"));

    cx = upcontext(aTHX_ uplevel, &cop, &ccstack, &cxix_from, &cxix_to);
    debug_print(("** cxix = (%d,%d)\n", cxix_from, cxix_to));
    if (cop == 0) {
	   debug_print(("**Setting cop to PL_curcop\n"));
	   cop = PL_curcop;
	}
	debug_print(("**Cop file = %s\n", CopFILE(cop)));

    context_vars(cx, ret, cop->cop_seq, PL_main_cv);

    for (i = cxix_from-1; i > cxix_to; --i) {
        debug_print(("** CxTYPE = %s (cxix = %d)\n",
            cxtype_name(CxTYPE(&ccstack[i])), i));
        switch (CxTYPE(&ccstack[i])) {
        case CXt_EVAL:
            switch(ccstack[i].blk_eval.old_op_type) {
            case OP_ENTEREVAL:
				context_vars(0, ret, ccstack[i].blk_oldcop->cop_seq,
								ccstack[i].blk_eval.cv);
				context_vars(0, ret, cop->cop_seq, ccstack[i].blk_eval.cv);
				break;
            case OP_REQUIRE:
				context_vars(0, ret, cop->cop_seq, ccstack[i].blk_eval.cv);
                goto END;
            }
            break;

        case CXt_SUB:
#ifdef CXt_FORMAT
        case CXt_FORMAT:
#endif
		    Perl_die(aTHX_ "PadWalker: internal error");
			exit(EXIT_FAILURE);
        }
    }

 END:
    EXTEND(SP, 1);
    PUSHs(sv_2mortal(newRV_noinc((SV*)ret)));
#line 308 "PadWalker.c"
	PUTBACK;
	return;
    }
}

XS(XS_PadWalker_peek_sub); /* prototype to pass -Wmissing-prototypes */
XS(XS_PadWalker_peek_sub)
{
    dXSARGS;
    if (items != 1)
	Perl_croak(aTHX_ "Usage: PadWalker::peek_sub(cur_sv)");
    SP -= items;
    {
	SV*	cur_sv = ST(0);
#line 298 "PadWalker.xs"
    CV* cur_cv = (CV*)SvRV(cur_sv);
    HV* ret = newHV();
    AV* cv_padlist;
#line 327 "PadWalker.c"
#line 302 "PadWalker.xs"
    padlist_into_hash(CvPADLIST(cur_cv), ret, 0, CvDEPTH(cur_cv));
    EXTEND(SP, 1);
    PUSHs(sv_2mortal(newRV_noinc((SV*)ret)));
#line 332 "PadWalker.c"
	PUTBACK;
	return;
    }
}

XS(XS_PadWalker__upcontext); /* prototype to pass -Wmissing-prototypes */
XS(XS_PadWalker__upcontext)
{
    dXSARGS;
    if (items != 1)
	Perl_croak(aTHX_ "Usage: PadWalker::_upcontext(uplevel)");
    SP -= items;
    {
	I32	uplevel = (I32)SvIV(ST(0));
#line 310 "PadWalker.xs"
    /* I'm not sure why this is here, but I'll leave it in case
	 * somebody is using it in an insanely evil way. */
    XPUSHs(sv_2mortal(newSViv((U32)upcontext(aTHX_ uplevel, 0, 0, 0, 0))));
#line 351 "PadWalker.c"
	PUTBACK;
	return;
    }
}

#ifdef __cplusplus
extern "C"
#endif
XS(boot_PadWalker); /* prototype to pass -Wmissing-prototypes */
XS(boot_PadWalker)
{
    dXSARGS;
    char* file = __FILE__;

    XS_VERSION_BOOTCHECK ;

        newXS("PadWalker::peek_my", XS_PadWalker_peek_my, file);
        newXS("PadWalker::peek_sub", XS_PadWalker_peek_sub, file);
        newXS("PadWalker::_upcontext", XS_PadWalker__upcontext, file);
    XSRETURN_YES;
}



PadWalker.pm
package PadWalker;

use strict;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);

require Exporter;
require DynaLoader;

require 5.008;

@ISA = qw(Exporter DynaLoader);
@EXPORT_OK = qw(peek_my peek_our closed_over peek_sub var_name set_closed_over);
%EXPORT_TAGS = (all => \@EXPORT_OK);

$VERSION = '1.91';

bootstrap PadWalker $VERSION;

sub peek_my;
sub peek_our;
sub closed_over;
sub peek_sub;
sub var_name;

1;
__END__

=head1 NAME

PadWalker - play with other peoples' lexical variables

=head1 SYNOPSIS

  use PadWalker qw(peek_my peek_our peek_sub closed_over);
  ...

=head1 DESCRIPTION

PadWalker is a module which allows you to inspect (and even change!)
lexical variables in any subroutine which called you. It will only
show those variables which are in scope at the point of the call.

PadWalker is particularly useful for debugging. It's even
used by Perl's built-in debugger. (It can also be used
for evil, of course.)

I wouldn't recommend using PadWalker directly in production
code, but it's your call. Some of the modules that use
PadWalker internally are certainly safe for and useful
in production.

=over 4

=item peek_my LEVEL

=item peek_our LEVEL

The LEVEL argument is interpreted just like the argument to C<caller>.
So C<peek_my(0)> returns a reference to a hash of all the C<my>
variables that are currently in scope;
C<peek_my(1)> returns a reference to a hash of all the C<my>
variables that are in scope at the point where the current
sub was called, and so on.

C<peek_our> works in the same way, except that it lists
the C<our> variables rather than the C<my> variables.

The hash associates each variable name with a reference
to its value. The variable names include the sigil, so
the variable $x is represented by the string '$x'.

For example:

  my $x = 12;
  my $h = peek_my (0);
  ${$h->{'$x'}}++;

  print $x;  # prints 13

Or a more complex example:

  sub increment_my_x {
    my $h = peek_my (1);
    ${$h->{'$x'}}++;
  }

  my $x=5;
  increment_my_x;
  print $x;  # prints 6

=item peek_sub SUB

The C<peek_sub> routine takes a coderef as its argument, and returns a hash
of the C<my> variables used in that sub. The values will usually be undefined
unless the sub is in use (i.e. in the call-chain) at the time. On the other
hand:

  my $x = "Hello!";
  my $r = peek_sub(sub {$x})->{'$x'};
  print "$$r\n";	# prints 'Hello!'

If the sub defines several C<my> variables with the same name, you'll get the
last one. I don't know of any use for C<peek_sub> that isn't broken as a result
of this, and it will probably be deprecated in a future version in favour of
some alternative interface.

=item closed_over SUB

C<closed_over> is similar to C<peek_sub>, except that it only lists
the C<my> variables which are used in the subroutine but defined outside:
in other words, the variables which it closes over. This I<does> have
reasonable uses: see L<Data::Dump::Streamer>, for example (a future version
of which may in fact use C<closed_over>).

=item set_closed_over SUB, HASH_REF

C<set_closed_over> reassigns the pad variables that are closed over by the subroutine.

The second argument is a hash of references, much like the one returned from C<closed_over>.

=item var_name LEVEL, VAR_REF

=item var_name SUB,   VAR_REF

C<var_name(sub, var_ref)> returns the name of the variable referred to
by C<var_ref>, provided it is a C<my> variable used in the sub. The C<sub>
parameter can be either a CODE reference or a number. If it's a number,
it's treated the same way as the argument to C<peek_my>.

For example,

  my $foo;
  print var_name(0, \$foo);    # prints '$foo'
  
  sub my_name {
    return var_name(1, shift);
  }
  print my_name(\$foo);        # ditto

=back

=head1 AUTHOR

Robin Houston <robin@cpan.org>

With contributions from Richard Soberberg, Jesse Luehrs and
Yuval Kogman, bug-spotting from Peter Scott, Dave Mitchell and
Goro Fuji, and suggestions from demerphq.

=head1 SEE ALSO

Devel::LexAlias, Devel::Caller, Sub::Parameters

=head1 COPYRIGHT

Copyright (c) 2000-2009, Robin Houston. All Rights Reserved.
This module is free software. It may be used, redistributed
and/or modified under the same terms as Perl itself.

=cut
